README.md

```diff
@@ -0,0 +1,191 @@
+The @PostConstruct annotation defines startup behavior in a thread-safe manner. 
+For example, in `NumaflowSink` to start the server and in `EventPublisher` to initialize and validate the state of `PulsarTemplate`, ensuring that components are ready for use upon initialization.
+
+While using Spring Boot for this project provided lots of benefits there were also tradeoffs compared to if we just used native Java. 
+These include larger memory consumption, learning curve of spring boot, and challenges with
+debugging issues because of the layers of abstraction and the magic of auto-configuration, making it sometimes hard to trace and fix bug.
+
+Nonetheless, the benefits Spring Boot provided outweighed its cons, and was used for the project. 
+
+## Challenges / Errors 
+
+
+## "this.publisher" is null
+`java.lang.NullPointerException: Cannot invoke "com.jacque1ine.producer.EventPublisher.publishPlainMessage(String)" because "this.publisher" is null
+`
+
+**DIDNT WORK -**
+**Manually instantiated instance (`new NumaflowSink()`):**
+
+Was previously creating a new server like this: `server = new Server(new NumaflowSink());`
+The server class needs to be passed a Sinker to be created.
+
+This instance is manually created and not managed by Spring.
+Since it is not managed by Spring, it won't have any of its dependencies (like EventPublisher) injected. 
+So, the publisher field remains null, leading to the `NullPointerException`.
+
+**WORKED - Spring-managed instance (this)** 
+
+In this approach, this refers to the current instance of NumaflowSink that has been created and managed by Spring. Thus, all dependencies marked with @Autowired inside this instance have already been injected by Spring. Therefore, when this instance (this) is passed to the Server constructor, the publisher field is not null, and it works as expected.The @PostConstruct annotated startServer method will be invoked after the Spring context is fully initialized, ensuring all dependencies are injected.
+`server = new Server(this);` passes the fully-initialized Spring-managed NumaflowSink instance into the Server constructor.
+
+In general:
+
+**Spring-managed Beans**
+* When you let Spring manage a bean, it takes care of creating an instance of the bean, initializing it, and injecting any dependencies that it requires. 
+* Allows Spring to properly inject any other Spring-managed components into that bean, including fields annotated with @Autowired.
+
+
+**Manually Instantiated Beans**
+
+When you manually instantiate a bean using new, you bypass the entire Spring lifecycle. This means that Spring doesn’t get a chance to inject its dependencies into that instance.
+
+## Extra_hosts configuration: host.docker.internal 
+**Client**: requests resources or services from a server. 
+**Server**: provides resources or services to users
+
+In this project, the Numaflow source is the server, while Pulsar is the client. 
+Each Numaflow vertex is running as a seperate pod in Kubernetes. 
```

Comment:
A vertex is a unit which can scale up and down, meaning it can have more than one pod.